// devextreme/cjs/__internal/grids/tree_list/module_validating.js
/**
 * DevExtreme (cjs/__internal/grids/tree_list/module_validating.js)
 * Version: 22.2.6
 * Build date: Tue Apr 18 2023
 *
 * Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
var __importDefault = this && this.__importDefault || function(mod) {
    return mod && mod.__esModule ? mod : {
        default: mod
    }
};
Object.defineProperty(exports, "__esModule", {
    value: true
});
var extend_1 = require("../../../core/utils/extend");
var ui_grid_core_validating_1 = require("../../../ui/grid_core/ui.grid_core.validating");
var module_core_1 = __importDefault(require("./module_core"));
var EditingControllerExtender = extend_1.extend({}, ui_grid_core_validating_1.validatingModule.extenders.controllers.editing);
delete EditingControllerExtender.processItems;
delete EditingControllerExtender.processDataItem;
module_core_1.default.registerModule("validating", {
    defaultOptions: ui_grid_core_validating_1.validatingModule.defaultOptions,
    controllers: ui_grid_core_validating_1.validatingModule.controllers,
    extenders: {
        controllers: {
            editing: EditingControllerExtender,
            editorFactory: ui_grid_core_validating_1.validatingModule.extenders.controllers.editorFactory
        },
        views: ui_grid_core_validating_1.validatingModule.extenders.views
    }
});
