// devextreme/bundles/__internal/grids/tree_list/module_grid_view.js
/**
 * DevExtreme (bundles/__internal/grids/tree_list/module_grid_view.js)
 * Version: 22.2.6
 * Build date: Tue Apr 18 2023
 *
 * Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
var __importDefault = this && this.__importDefault || function(mod) {
    return mod && mod.__esModule ? mod : {
        default: mod
    }
};
Object.defineProperty(exports, "__esModule", {
    value: true
});
var ui_grid_core_grid_view_1 = require("../../../ui/grid_core/ui.grid_core.grid_view");
var module_core_1 = __importDefault(require("./module_core"));
var GridView = ui_grid_core_grid_view_1.gridViewModule.views.gridView.inherit({
    _getWidgetAriaLabel: function() {
        return "dxTreeList-ariaTreeList"
    },
    _getTableRoleName: function() {
        return "treegrid"
    }
});
module_core_1.default.registerModule("gridView", {
    defaultOptions: ui_grid_core_grid_view_1.gridViewModule.defaultOptions,
    controllers: ui_grid_core_grid_view_1.gridViewModule.controllers,
    views: {
        gridView: GridView
    },
    extenders: {
        controllers: {
            resizing: {
                _toggleBestFitMode: function(isBestFit) {
                    this.callBase(isBestFit);
                    var $rowsTable = this._rowsView.getTableElement();
                    $rowsTable.find(".dx-treelist-cell-expandable").toggleClass(this.addWidgetPrefix("best-fit"), isBestFit)
                }
            }
        }
    }
});
