// devextreme/bundles/__internal/grids/pivot_grid/field_chooser/module.js
/**
 * DevExtreme (bundles/__internal/grids/pivot_grid/field_chooser/module.js)
 * Version: 22.2.6
 * Build date: Tue Apr 18 2023
 *
 * Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
var __importDefault = this && this.__importDefault || function(mod) {
    return mod && mod.__esModule ? mod : {
        default: mod
    }
};
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.FieldChooser = void 0;
var renderer_1 = __importDefault(require("../../../../core/renderer"));
var icon_1 = require("../../../../core/utils/icon");
var window_1 = require("../../../../core/utils/window");
var type_1 = require("../../../../core/utils/type");
var extend_1 = require("../../../../core/utils/extend");
var iterator_1 = require("../../../../core/utils/iterator");
var message_1 = __importDefault(require("../../../../localization/message"));
var component_registrator_1 = __importDefault(require("../../../../core/component_registrator"));
var tree_view_1 = __importDefault(require("../../../../ui/tree_view"));
var context_menu_1 = __importDefault(require("../../../../ui/context_menu"));
var module_widget_utils_1 = require("../module_widget_utils");
var module_base_1 = require("./module_base");
require("../data_source/module");
var const_1 = require("./const");
var DIV = "<div>";
var hasWindow = window_1.hasWindow();

function getDimensionFields(item, fields) {
    var result = [];
    if (item.items) {
        for (var i = 0; i < item.items.length; i += 1) {
            result.push.apply(result, getDimensionFields(item.items[i], fields))
        }
    } else if (type_1.isDefined(item.index)) {
        result.push(fields[item.index])
    }
    return result
}

function getFirstItem(item, condition) {
    if (item.items) {
        for (var i = 0; i < item.items.length; i += 1) {
            var childrenItem = getFirstItem(item.items[i], condition);
            if (childrenItem) {
                return childrenItem
            }
        }
    }
    if (condition(item)) {
        return item
    }
    return
}
var compareOrder = [function(a, b) {
    var aValue = -!!a.isMeasure;
    var bValue = +!!b.isMeasure;
    return aValue + bValue
}, function(a, b) {
    var aValue = -!!(a.items && a.items.length);
    var bValue = +!!(b.items && b.items.length);
    return aValue + bValue
}, function(a, b) {
    var aValue = +!!(false === a.isMeasure && a.field && a.field.levels && a.field.levels.length);
    var bValue = -!!(false === b.isMeasure && b.field && b.field.levels && b.field.levels.length);
    return aValue + bValue
}, module_widget_utils_1.getCompareFunction((function(item) {
    return item.text
}))];

function compareItems(a, b) {
    var result = 0;
    var i = 0;
    while (!result && compareOrder[i]) {
        result = compareOrder[i++](a, b)
    }
    return result
}

function getScrollable(container) {
    return container.find("." + const_1.CLASSES.scrollable.self).dxScrollable("instance")
}
var FieldChooser = module_base_1.FieldChooserBase.inherit({
    _getDefaultOptions: function() {
        return extend_1.extend(this.callBase(), {
            height: 400,
            layout: 0,
            dataSource: null,
            encodeHtml: true,
            onContextMenuPreparing: null,
            allowSearch: false,
            searchTimeout: 500,
            texts: {
                columnFields: message_1.default.format("dxPivotGrid-columnFields"),
                rowFields: message_1.default.format("dxPivotGrid-rowFields"),
                dataFields: message_1.default.format("dxPivotGrid-dataFields"),
                filterFields: message_1.default.format("dxPivotGrid-filterFields"),
                allFields: message_1.default.format("dxPivotGrid-allFields")
            }
        })
    },
    _refreshDataSource: function() {
        var that = this;
        that._expandedPaths = [];
        that._changedHandler = that._changedHandler || function() {
            iterator_1.each(that._dataChangedHandlers, (function(_, func) {
                func()
            }));
            that._fireContentReadyAction();
            that._skipStateChange = true;
            that.option("state", that._dataSource.state());
            that._skipStateChange = false
        };
        that._disposeDataSource();
        that.callBase();
        that._dataSource && that._dataSource.on("changed", that._changedHandler)
    },
    _disposeDataSource: function() {
        var dataSource = this._dataSource;
        if (dataSource) {
            dataSource.off("changed", this._changedHandler);
            this._dataSource = void 0
        }
    },
    _dispose: function() {
        this._disposeDataSource();
        this.callBase.apply(this, arguments)
    },
    _init: function() {
        this.callBase();
        this._refreshDataSource();
        this._dataChangedHandlers = [];
        this._initActions()
    },
    _initActions: function() {
        this._actions = {
            onContextMenuPreparing: this._createActionByOption("onContextMenuPreparing")
        }
    },
    _trigger: function(eventName, eventArg) {
        this._actions[eventName](eventArg)
    },
    _setOptionsByReference: function() {
        this.callBase();
        extend_1.extend(this._optionsByReference, {
            dataSource: true
        })
    },
    _optionChanged: function(args) {
        switch (args.name) {
            case "dataSource":
                this._refreshDataSource();
                this._invalidate();
                break;
            case "layout":
            case "texts":
            case "allowSearch":
            case "searchTimeout":
            case "encodeHtml":
                this._invalidate();
                break;
            case "onContextMenuPreparing":
                this._actions[args.name] = this._createActionByOption(args.name);
                break;
            default:
                this.callBase(args)
        }
    },
    _clean: function(skipStateSetting) {
        !skipStateSetting && this._dataSource && this.option("state", this._dataSource.state());
        this.$element().children("." + const_1.CLASSES.fieldChooser.container).remove()
    },
    _renderLayout0: function($container) {
        $container.addClass(const_1.CLASSES.layout.zero);
        var $row1 = renderer_1.default(DIV).addClass(const_1.CLASSES.row).appendTo($container);
        var $row2 = renderer_1.default(DIV).addClass(const_1.CLASSES.row).appendTo($container);
        var $col1 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($row1);
        var $col2 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($row1);
        var $col3 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($row2);
        var $col4 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($row2);
        this._renderArea($col1, "all");
        this._renderArea($col2, "row");
        this._renderArea($col2, "column");
        this._renderArea($col3, "filter");
        this._renderArea($col4, "data")
    },
    _renderLayout1: function($container) {
        var $col1 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($container);
        var $col2 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($container);
        this._renderArea($col1, "all");
        this._renderArea($col2, "filter");
        this._renderArea($col2, "row");
        this._renderArea($col2, "column");
        this._renderArea($col2, "data")
    },
    _renderLayout2: function($container) {
        $container.addClass(const_1.CLASSES.layout.second);
        var $row1 = renderer_1.default(DIV).addClass(const_1.CLASSES.row).appendTo($container);
        this._renderArea($row1, "all");
        var $row2 = renderer_1.default(DIV).addClass(const_1.CLASSES.row).appendTo($container);
        var $col1 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($row2);
        var $col2 = renderer_1.default(DIV).addClass(const_1.CLASSES.col).appendTo($row2);
        this._renderArea($col1, "filter");
        this._renderArea($col1, "row");
        this._renderArea($col2, "column");
        this._renderArea($col2, "data")
    },
    _initMarkup: function() {
        var $element = this.$element();
        var $container = renderer_1.default(DIV).addClass(const_1.CLASSES.fieldChooser.container).appendTo($element);
        var layout = this.option("layout");
        this.callBase();
        $element.addClass(const_1.CLASSES.fieldChooser.self).addClass(const_1.CLASSES.pivotGrid.fieldsContainer);
        this._dataChangedHandlers = [];
        var dataSource = this._dataSource;
        var currentState = "instantly" !== this.option("applyChangesMode") && dataSource && dataSource.state();
        currentState && this.option("state") && dataSource.state(this.option("state"), true);
        if (0 === layout) {
            this._renderLayout0($container)
        } else if (1 === layout) {
            this._renderLayout1($container)
        } else {
            this._renderLayout2($container)
        }
        currentState && dataSource.state(currentState, true)
    },
    _renderContentImpl: function() {
        this.callBase();
        this.renderSortable();
        this._renderContextMenu();
        this.updateDimensions()
    },
    _fireContentReadyAction: function() {
        if (!this._dataSource || !this._dataSource.isLoading()) {
            this.callBase()
        }
    },
    _getContextMenuArgs: function(dxEvent) {
        var targetFieldElement = renderer_1.default(dxEvent.target).closest("." + const_1.CLASSES.area.field);
        var targetGroupElement = renderer_1.default(dxEvent.target).closest("." + const_1.CLASSES.area.fieldList);
        var field;
        var area;
        if (targetFieldElement.length) {
            var fieldCopy = targetFieldElement.data("field");
            if (fieldCopy) {
                field = this.getDataSource().field(fieldCopy.index) || fieldCopy
            }
        }
        if (targetGroupElement.length) {
            area = targetGroupElement.attr("group")
        }
        return {
            event: dxEvent,
            field: field,
            area: area,
            items: []
        }
    },
    _renderContextMenu: function() {
        var that = this;
        var $container = that.$element();
        if (that._contextMenu) {
            that._contextMenu.$element().remove()
        }
        that._contextMenu = that._createComponent(renderer_1.default(DIV).appendTo($container), context_menu_1.default, {
            onPositioning: function(actionArgs) {
                var event = actionArgs.event;
                if (!event) {
                    return
                }
                var args = that._getContextMenuArgs(event);
                that._trigger("onContextMenuPreparing", args);
                if (args.items && args.items.length) {
                    actionArgs.component.option("items", args.items)
                } else {
                    actionArgs.cancel = true
                }
            },
            target: $container,
            onItemClick: function(params) {
                params.itemData.onItemClick && params.itemData.onItemClick(params)
            },
            cssClass: const_1.CLASSES.fieldChooser.contextMenu
        })
    },
    _createTreeItems: function(fields, groupFieldNames, path) {
        var that = this;
        var isMeasure;
        var resultItems = [];
        var groupedItems = [];
        var groupFieldName = groupFieldNames[0];
        var fieldsByGroup = {};
        if (!groupFieldName) {
            iterator_1.each(fields, (function(_, field) {
                var icon;
                if (true === field.isMeasure) {
                    icon = const_1.ICONS.measure
                }
                if (false === field.isMeasure) {
                    icon = field.groupName ? const_1.ICONS.hierarchy : const_1.ICONS.dimension
                }
                resultItems.push({
                    index: field.index,
                    field: field,
                    key: field.dataField,
                    selected: type_1.isDefined(field.area),
                    text: field.caption || field.dataField,
                    icon: icon,
                    isMeasure: field.isMeasure,
                    isDefault: field.isDefault
                })
            }))
        } else {
            iterator_1.each(fields, (function(_, field) {
                var groupName = field[groupFieldName] || "";
                fieldsByGroup[groupName] = fieldsByGroup[groupName] || [];
                fieldsByGroup[groupName].push(field);
                if (void 0 === isMeasure) {
                    isMeasure = true
                }
                isMeasure = isMeasure && true === field.isMeasure
            }));
            iterator_1.each(fieldsByGroup, (function(groupName, fields) {
                var currentPath = path ? path + "." + groupName : groupName;
                var items = that._createTreeItems(fields, groupFieldNames.slice(1), currentPath);
                if (groupName) {
                    groupedItems.push({
                        key: groupName,
                        text: groupName,
                        path: currentPath,
                        isMeasure: items.isMeasure,
                        expanded: that._expandedPaths.includes(currentPath),
                        items: items
                    })
                } else {
                    resultItems = items
                }
            }));
            resultItems = groupedItems.concat(resultItems);
            resultItems.isMeasure = isMeasure
        }
        return resultItems
    },
    _createFieldsDataSource: function(dataSource) {
        var fields = dataSource && dataSource.fields() || [];
        fields = fields.filter((function(field) {
            return false !== field.visible && !type_1.isDefined(field.groupIndex)
        }));
        var treeItems = this._createTreeItems(fields, ["dimension", "displayFolder"]);
        module_widget_utils_1.foreachDataLevel(treeItems, (function(items) {
            items.sort(compareItems)
        }), 0, "items");
        return treeItems
    },
    _renderFieldsTreeView: function(container) {
        var that = this;
        var dataSource = that._dataSource;
        var treeView = that._createComponent(container, tree_view_1.default, {
            dataSource: that._createFieldsDataSource(dataSource),
            showCheckBoxesMode: "normal",
            expandNodesRecursive: false,
            searchEnabled: that.option("allowSearch"),
            searchTimeout: that.option("searchTimeout"),
            useNativeScrolling: false,
            itemTemplate: function(itemData, itemIndex, itemElement) {
                var _a;
                var $item = renderer_1.default("<div>").toggleClass(const_1.CLASSES.area.field, !itemData.items).attr(const_1.ATTRIBUTES.treeViewItem, true).data("field", itemData.field).appendTo(itemElement);
                if (itemData.icon) {
                    null === (_a = icon_1.getImageContainer(itemData.icon)) || void 0 === _a ? void 0 : _a.appendTo($item)
                }
                renderer_1.default("<span>").text(itemData.text).appendTo($item)
            },
            onItemCollapsed: function(e) {
                var index = that._expandedPaths.indexOf(e.itemData.path);
                if (index >= 0) {
                    that._expandedPaths.splice(index, 1)
                }
            },
            onItemExpanded: function(e) {
                var index = that._expandedPaths.indexOf(e.itemData.path);
                if (index < 0) {
                    that._expandedPaths.push(e.itemData.path)
                }
            },
            onItemSelectionChanged: function(e) {
                var data = e.itemData;
                var field;
                var fields;
                var needSelectDefaultItem = true;
                var area;
                if (data.items) {
                    if (data.selected) {
                        treeView.unselectItem(data);
                        return
                    }
                    that._processDemandState((function() {
                        fields = getDimensionFields(data, dataSource.fields());
                        for (var i = 0; i < fields.length; i += 1) {
                            if (fields[i].area) {
                                needSelectDefaultItem = false;
                                break
                            }
                        }
                    }));
                    if (needSelectDefaultItem) {
                        var item = getFirstItem(data, (function(item) {
                            return item.isDefault
                        })) || getFirstItem(data, (function(item) {
                            return type_1.isDefined(item.index)
                        }));
                        item && treeView.selectItem(item);
                        return
                    }
                } else {
                    field = dataSource.fields()[data.index];
                    if (data.selected) {
                        area = field.isMeasure ? "data" : "column"
                    }
                    if (field) {
                        fields = [field]
                    }
                }
                that._applyChanges(fields, {
                    area: area,
                    areaIndex: void 0
                })
            }
        });
        that._dataChangedHandlers.push((function() {
            var scrollable = getScrollable(container);
            var scrollTop = scrollable ? scrollable.scrollTop() : 0;
            treeView.option({
                dataSource: that._createFieldsDataSource(dataSource)
            });
            scrollable = getScrollable(container);
            if (scrollable) {
                scrollable.scrollTo({
                    y: scrollTop
                });
                scrollable.update()
            }
        }))
    },
    _renderAreaFields: function($container, area) {
        var that = this;
        var dataSource = that._dataSource;
        var fields = dataSource ? extend_1.extend(true, [], dataSource.getAreaFields(area, true)) : [];
        $container.empty();
        iterator_1.each(fields, (function(_, field) {
            if (false !== field.visible) {
                that.renderField(field, true).appendTo($container)
            }
        }))
    },
    _renderArea: function(container, area) {
        var that = this;
        var $areaContainer = renderer_1.default(DIV).addClass(const_1.CLASSES.area.self).appendTo(container);
        var $fieldsHeaderContainer = renderer_1.default(DIV).addClass(const_1.CLASSES.area.fieldListHeader).appendTo($areaContainer);
        var caption = that.option("texts." + area + "Fields");
        var $fieldsContent;
        var render;
        renderer_1.default("<span>").addClass(const_1.CLASSES.area.icon).addClass("dx-area-icon-" + area).appendTo($fieldsHeaderContainer);
        renderer_1.default("<span>").html("&nbsp;").appendTo($fieldsHeaderContainer);
        renderer_1.default("<span>").addClass(const_1.CLASSES.area.caption).text(caption).appendTo($fieldsHeaderContainer);
        var $fieldsContainer = renderer_1.default(DIV).addClass(const_1.CLASSES.area.fieldList).addClass(const_1.CLASSES.pivotGrid.dragAction).appendTo($areaContainer);
        if ("all" !== area) {
            $fieldsContainer.attr("group", area).attr(const_1.ATTRIBUTES.allowScrolling, true);
            $fieldsContent = renderer_1.default(DIV).addClass(const_1.CLASSES.area.fieldContainer).appendTo($fieldsContainer);
            render = function() {
                that._renderAreaFields($fieldsContent, area)
            };
            that._dataChangedHandlers.push(render);
            render();
            $fieldsContainer.dxScrollable({
                useNative: false
            })
        } else {
            $areaContainer.addClass(const_1.CLASSES.allFields);
            $fieldsContainer.addClass(const_1.CLASSES.treeView.borderVisible);
            that._renderFieldsTreeView($fieldsContainer)
        }
    },
    _getSortableOptions: function() {
        return {}
    },
    _adjustSortableOnChangedArgs: function() {},
    resetTreeView: function() {
        var treeView = this.$element().find("." + const_1.CLASSES.treeView.self).dxTreeView("instance");
        if (treeView) {
            treeView.option("searchValue", "");
            treeView.collapseAll()
        }
    },
    applyChanges: function() {
        var state = this.option("state");
        if (type_1.isDefined(state)) {
            this._dataSource.state(state)
        }
    },
    cancelChanges: function() {
        var dataSource = this._dataSource;
        if (!dataSource.isLoading()) {
            this.option("state", dataSource.state());
            return true
        }
        return false
    },
    getDataSource: function() {
        return this._dataSource
    },
    updateDimensions: function() {
        var $scrollableElements = this.$element().find("." + const_1.CLASSES.area.self + " ." + const_1.CLASSES.scrollable.self);
        $scrollableElements.dxScrollable("update")
    },
    _visibilityChanged: function(visible) {
        if (visible && hasWindow) {
            this.updateDimensions()
        }
    }
});
exports.FieldChooser = FieldChooser;
component_registrator_1.default("dxPivotGridFieldChooser", FieldChooser);
exports.default = {
    FieldChooser: FieldChooser
};
